### PR TITLE
IRGen: get/await_async_continuation support

### PR DESCRIPTION
**Pull request by aschwaighofer**
rdar://71124933

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci please test

File: lib/IRGen/IRGenFunction.cpp [hunk 1]
```diff
@@ -513,3 +513,249 @@ llvm::Value *IRGenFunction::alignUpToMaximumAlignment(llvm::Type *sizeTy, llvm::
...
+  // remaining control flow).
+  auto contAwaitSyncAddr =
+      Builder.CreateStructGEP(AsyncCoroutineCurrentContinuationContext, 1);
+  auto null = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 0);
+  auto one = llvm::ConstantInt::get(
+      contAwaitSyncAddr->getType()->getPointerElementType(), 1);
+  auto results = Builder.CreateAtomicCmpXchg(
+      contAwaitSyncAddr, null, one,
+      llvm::AtomicOrdering::AcquireRelease /*success ordering*/,
```
Comment (aschwaighofer) [inline lib/IRGen/IRGenFunction.cpp:677]: The ordering needs to be swapped, we continue with control flow if the swap failed (i.e all information is available: both the await point was  reached and the continuation was resumed).

File: lib/IRGen/IRGenFunction.cpp [hunk 2]
```diff
@@ -513,3 +513,249 @@ llvm::Value *IRGenFunction::alignUpToMaximumAlignment(llvm::Type *sizeTy, llvm::
...
+    arguments.push_back(
+        Builder.CreateBitOrPointerCast(getAsyncExecutor(), IGM.Int8PtrTy));
+    arguments.push_back(Builder.CreateBitOrPointerCast(
+        AsyncCoroutineCurrentContinuationContext, IGM.Int8PtrTy));
+    auto *id = Builder.CreateIntrinsicCall(llvm::Intrinsic::coro_suspend_async,
+
+                                           arguments);
+    auto results = Builder.CreateAtomicCmpXchg(
+        contAwaitSyncAddr, null, one,
+        llvm::AtomicOrdering::AcquireRelease /*success ordering*/,
```
Comment (aschwaighofer) [inline lib/IRGen/IRGenFunction.cpp:714]: The ordering needs to be swapped, we continue with control flow if the swap failed (i.e all information is available: both the await point was reached and the continuation was resumed).

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

File: lib/IRGen/IRGenFunction.cpp [hunk 3]
```diff
@@ -513,3 +514,253 @@ llvm::Value *IRGenFunction::alignUpToMaximumAlignment(llvm::Type *sizeTy, llvm::
...
+  //
+  // Where:
+  //
+  // struct {
+  //   AsyncContext *resumeCtxt;
+  //   void *awaitSynchronization;
+  //   SwiftError *errResult;
+  //   union {
+  //     IndirectResult *result;
+  //     DirectResult *result;
```
Comment (jckarter) [inline lib/IRGen/IRGenFunction.cpp:568]: What's the difference between IndirectResult and DirectResult?

Comment (aschwaighofer) [reply lib/IRGen/IRGenFunction.cpp:568]: Nothing, really.

Sometimes, we store direct results directly in a data structure.

In such a case I would have written:
```
struct {
   //   AsyncContext *resumeCtxt;
   //   void *awaitSynchronization;
   //   SwiftError *errResult;
   //   union {
   //     IndirectResult *result;
   //     DirectResult result;
```

I wanted to point out that the storage for both cases is the same.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/IRGen/IRGenFunction.cpp (hunk 3 above)
Comment (jckarter) [reply lib/IRGen/IRGenFunction.cpp:568]: I see, so for a word-sized result, we store it directly in the context? That makes sense.

Comment (aschwaighofer) [reply lib/IRGen/IRGenFunction.cpp:568]: No we don't currently.

With the sentence:

>Sometimes, we store direct results directly in a data structure.

I wanted to say: In other places such as the context passed to async functions (not the continuation context mentioned here) we store direct results in place and indirect results by address:
https://github.com/apple/swift/blob/main/lib/IRGen/GenCall.cpp#L161
https://github.com/apple/swift/blob/main/lib/IRGen/GenCall.cpp#L190

The notation:
```
  //   union {
  //     IndirectResult *result;
  //     DirectResult *result;
```
is meant to convey that both direct and indirect results are stored the same (as addresses to memory that stores the result).

**Comment by aschwaighofer** [comment]
@swift-ci please test

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

File: lib/IRGen/IRGenFunction.cpp (hunk 3 above)
Comment (rjmccall) [reply lib/IRGen/IRGenFunction.cpp:568]: On net, I think just saying `Result*` would probably be clearer.

Comment (aschwaighofer) [reply lib/IRGen/IRGenFunction.cpp:568]: https://github.com/apple/swift/pull/34937

**Pull request merged**